### ISSUE TITLE
Adding command into recipe's content of stack causes issue when starting workspace

### ISSUE THREAD

**Issue by vikas-tawniya** [eclipse-che/che#12844]
### Description
If you add a command in the recipe content like following:

 "recipe": {
              "type": "kubernetes",
              "content": "kind: List\nitems:\n - \n  apiVersion: v1\n  kind: Pod\n  metadata:\n   name: hub-tooling\n  spec:\n   initContainers:\n    - \n     image: '192.168.2.100:30500/license'\n     name: attach-license\n     imagePullPolicy: Always\n     command: ['/bin/sh', '-c', 'cp /usr/xyz.xxx /usr/xyz.csa']\n     resources:\n      limits:\n       memory: 512Mi\n   containers:\n    - \n     image: 'wsskeleton/che-machine-exec:latest'\n     name: machine-exec\n    resources:\n      limits:\n       memory: 512Mi\n",
              "contentType": "text/x-yaml"
            } 

Stack gets created but the command [] array gets converted into list in the workspace config

### Reproduction Steps
Format might be wrong here by a space or two but the original one that I tried worked for creating a stack.
So, basically if you try adding a command to your container in the stack recipe and then check it in your stack raw configuration, it will appear fine having all square bracket and quotes and everything.
Then when you create a workspace with the same stack and check config of it, you will notice that the content of command in square bracket gets converted into list like : 

"command:\n      - /bin/sh\n      - '-c'\n      - cp /usr/xyz.xxx /usr/xyz.csa\n"

Now workspace doesn't start at all. 

When you try and open it it will keep loading forever and if you check for resources inside the workspace namespace, you won't find even a single resource.

Is it not possible to use command for a image inside recipe ?
Or am I doing it wrong ?

**Comment by VikTawniya** [comment]
Or is it something to do with initContainers ?

I tried testing workspace with same stack and blank project and below is what I could see in che logs:

2019-03-07 09:01:10,916[main]             [INFO ] [o.a.catalina.startup.Catalina 683]   - Server startup in 48196 ms
2019-03-07 09:05:34,925[nio-8080-exec-4]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 469]    - Workspace 'admin/test-wksp-k556' with id 'workspace0bezh6bhoqkxahyw' created by user 'admin'
2019-03-07 09:05:38,476[nio-8080-exec-4]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 322]   - Starting workspace 'admin/test-wksp-k556' with id 'workspace0bezh6bhoqkxahyw' by user 'admin'
2019-03-07 09:05:38,980[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 226]  - Failed to start Kubernetes runtime of workspace workspace0bezh6bhoqkxahyw. Cause: null
2019-03-07 09:05:38,988[aceSharedPool-0]  [WARN ] [i.f.k.c.i.VersionUsageUtils 55]      - The client is using resource type 'ingresses' with unstable version 'v1beta1'
2019-03-07 09:05:39,545[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 383]   - Workspace 'admin:test-wksp-k556' with id 'workspace0bezh6bhoqkxahyw' start failed
2019-03-07 09:05:39,548[aceSharedPool-0]  [ERROR] [o.e.c.a.w.s.WorkspaceRuntimes 393]   - null
org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException: null
        at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInternalRuntime.wrapAndRethrow(KubernetesInternalRuntime.java:710)
        at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInternalRuntime.internalStart(KubernetesInternalRuntime.java:248)
        at org.eclipse.che.api.workspace.server.spi.InternalRuntime.start(InternalRuntime.java:148)
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartRuntimeTask.run(WorkspaceRuntimes.java:359)
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:38)
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: null
        at org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy.provision(CommonPVCStrategy.java:123)
        at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesEnvironmentProvisioner$KubernetesEnvironmentProvisionerImpl.provision(KubernetesEnvironmentProvisioner.java:122)
        at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInternalRuntime.internalStart(KubernetesInternalRuntime.java:181)
        ... 7 common frames omitted

**Comment by VikTawniya** [comment]
This is what my actual recipe looks like:
```
kind: List
items:
 - 
  apiVersion: v1
  kind: Pod
  metadata:
   name: abc-tooling
  spec:
   volumes:
    - 
     name: cfg
     emptyDir: {}
   initContainers:
    - 
     image: '<docker_reg_url>/xyzFile'
     name: attach-xyzFile
     imagePullPolicy: Always
     command: ['/bin/sh', '-c', 'cp /xyzFileDir/xyzFile /cfg/xyzFile']
     resources:
      limits:
       memory: 512Mi
     volumeMounts:
      - 
       name: cfg
       mountPath: /cfg
   containers:
    - 
     image: '<docker_reg_url>/theia:latest'
     name: abc-theia-ide
     resources:
      limits:
       memory: 1024Mi
    - 
     image: '<docker_reg_url>/abc-tools:latest'
     name: abc-tools
     command: ['/bin/sh', '-c', 'cp /cfg/xyzFile /abc/def/xyzFile']
     volumeMounts:
      - 
       name: cfg
       mountPath: /cfg
     resources:
      limits:
       memory: 1024Mi
    - 
     image: 'wsskeleton/che-machine-exec:latest'
     name: machine-exec
     volumeMounts:
      - 
       name: dockersock
       mountPath: /var/run/docker.sock
     resources:
      limits:
       memory: 512Mi
```

**Comment by eivantsov** [comment]
@VikTawniya you can override command. Can you share your workspace configuration? It's config tab in workspace details page in Dashboard.

**Comment by VikTawniya** [comment]
> @VikTawniya you can override command. Can you share your workspace configuration? It's config tab in workspace details page in Dashboard.

Here you go
```
{
  "defaultEnv": "default",
  "environments": {
    "default": {
      "machines": {
        "abc-tooling/abc-theia-ide": {
          "attributes": {
            "memoryLimitBytes": "2147483648"
          },
          "servers": {
            "theia": {
              "attributes": {
                "type": "ide"
              },
              "port": "3000",
              "path": "/",
              "protocol": "http"
            },
            "theia-dev": {
              "attributes": {
                "type": "ide-dev"
              },
              "port": "3030",
              "protocol": "http"
            }
          },
          "volumes": {
            "projects": {
              "path": "/projects"
            }
          },
          "installers": [],
          "env": {
            "COMPILE_TOOLS_MACHINE_NAME": "abc-tooling/abc-tools",
            "USE_LOCAL_GIT": "true"
          }
        },
        "abc-tooling/machine-exec": {
          "attributes": {
            "memoryLimitBytes": "536870912"
          },
          "servers": {
            "machine-exec": {
              "attributes": {
                "type": "terminal"
              },
              "port": "4444",
              "protocol": "ws"
            }
          },
          "volumes": {},
          "installers": [],
          "env": {}
        },
        "abc-tooling/abc-tools": {
          "attributes": {
            "memoryLimitBytes": "1073741824"
          },
          "servers": {},
          "volumes": {
            "projects": {
              "path": "/projects"
            }
          },
          "installers": [],
          "env": {}
        }
      },
      "recipe": {
        "type": "kubernetes",
        "content": "kind: List\nitems:\n - \n  apiVersion: v1\n  kind: Pod\n  metadata:\n   name: abc-tooling\n  spec:\n   volumes:\n    - \n     name: cfg\n     emptyDir: {}\n   initContainers:\n    - \n     image: '<docker_reg_url>/xyzFile'\n     name: attach-xyzFile\n     imagePullPolicy: Always\n     command: ['/bin/sh', '-c', 'cp /xyzFileDir/xyzFile /cfg/xyzFile']\n     resources:\n      limits:\n       memory: 512Mi\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n   containers:\n    - \n     image: '<docker_reg_url>/theia:latest'\n     name: abc-theia-ide\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: '<docker_reg_url>/abc-tools:latest'\n     name: abc-tools\n     command: ['/bin/sh', '-c', 'cp /cfg/xyzFile /abc/def/xyzFile']\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: 'wsskeleton/che-machine-exec:latest'\n     name: machine-exec\n     volumeMounts:\n      - \n       name: dockersock\n       mountPath: /var/run/docker.sock\n     resources:\n      limits:\n       memory: 512Mi\n",
        "contentType": "text/x-yaml"
      }
    }
  },
  "projects": [],
  "name": "wksp-vwpm",
  "attributes": {},
  "commands": [],
  "links": []
}
```

I am not sure if I should add a machine for initContainers as well. If I do it complains about machine not existing in the recipe. Are InitContainers not allowed ?
Or we have to specify them in machines in some special way ?

**Comment by sleshchenko** [comment]
> I am not sure if I should add a machine for initContainers as well. If I do it complains about machine not existing in the recipe. 

You should not add machines for initContainers

> Are InitContainers not allowed ?

As far as I know, initContainers are allowed.

**Comment by eivantsov** [comment]
@VikTawniya it's not allowed to use hostPath mount. So, I am unable to create a workspace from config you have provided. 

Having removed that mount things look ok to me, though I cannot actually check it since you did not provide real images in the config.

**Comment by vtawniya** [comment]
> @VikTawniya it's not allowed to use hostPath mount. So, I am unable to create a workspace from config you have provided.
> 
> Having removed that mount things look ok to me, though I cannot actually check it since you did not provide real images in the config.

But I am not using any hostPath mounts it seems

**Comment by vtawniya** [comment]
> @VikTawniya it's not allowed to use hostPath mount. So, I am unable to create a workspace from config you have provided.
> 
> Having removed that mount things look ok to me, though I cannot actually check it since you did not provide real images in the config.

And also, I think hostPath mounts are also fine. Because I have used some recipes where I am mounting dockerSock

**Comment by eivantsov** [comment]
```
        "content": "kind: List\nitems:\n - \n  apiVersion: v1\n  kind: Pod\n  metadata:\n   name: abc-tooling\n  spec:\n   volumes:\n    - \n     name: cfg\n     emptyDir: {}\n   initContainers:\n    - \n     image: '<docker_reg_url>/xyzFile'\n     name: attach-xyzFile\n     imagePullPolicy: Always\n     command: ['/bin/sh', '-c', 'cp /xyzFileDir/xyzFile /cfg/xyzFile']\n     resources:\n      limits:\n       memory: 512Mi\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n   containers:\n    - \n     image: '<docker_reg_url>/theia:latest'\n     name: abc-theia-ide\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: '<docker_reg_url>/abc-tools:latest'\n     name: abc-tools\n     command: ['/bin/sh', '-c', 'cp /cfg/xyzFile /abc/def/xyzFile']\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: 'wsskeleton/che-machine-exec:latest'\n     name: machine-exec\n     volumeMounts:\n      - \n       name: dockersock\n       mountPath: /var/run/docker.sock\n     resources:\n      limits:\n       memory: 512Mi\n",
```

Here's a hostPath. I am checking on OpenShift though. Anyway, you do not need it at all for exec machine since it uses k8s API to run execs.

What do you want to accomplish with this workspace? Maybe, there's a cleaner way to create a start it.

**Comment by vtawniya** [comment]
@eivantsov 
I am trying to use a sidecar to copy license from one initcontainer to other container machine.

I was suppose to remove extra mounts like dockersock from the recipe to avoid confusion, but I forgot one it seems. :D

**Comment by vtawniya** [comment]
@eivantsov 
The workspace that I have works fine in itself. We are using theia ide and other machine for compiling and running projects.
But I was trying to introduce a simple initContainer to insert licenses required as sidecar and just copy it from there to other tooling machine.
But Something did not work fine in my case. Workspace itself never started with my changes. and logs showed nulls

**Comment by eivantsov** [comment]
I'd appreciate if you share a real config that I can try and reproduce the error, otherwise it's difficult to help

**Comment by vtawniya** [comment]
@eivantsov 
Here you go:

```
{
      "name": "Dummy",
      "components": [],
      "tags": [
        "Kubernetes"
      ],
      "id": "dummy-hub-stack",
      "workspaceConfig": {
        "defaultEnv": "default",
        "environments": {
          "default": {
            "machines": {
              "hub-tooling/hub-theia-ide": {
                "attributes": {
                  "memoryLimitBytes": "2147483648"
                },
                "servers": {
                  "theia": {
                    "attributes": {
                      "type": "ide"
                    },
                    "port": "3000",
                    "path": "/",
                    "protocol": "http"
                  },
                  "theia-dev": {
                    "attributes": {
                      "type": "ide-dev"
                    },
                    "port": "3030",
                    "protocol": "http"
                  }
                },
                "volumes": {
                  "projects": {
                    "path": "/projects"
                  }
                },
                "installers": [],
                "env": {
                   "USE_LOCAL_GIT":"true",
                   "COMPILE_TOOLS_MACHINE_NAME":"hub-tooling/dummy-tools"
                }
              },
              "hub-tooling/machine-exec": {
                "attributes": {
                  "memoryLimitBytes": "536870912"
                },
                "servers": {
                  "machine-exec": {
                    "attributes": {
                      "type": "terminal"
                    },
                    "port": "4444",
                    "protocol": "ws"
                  }
                },
                "volumes": {},
                "installers": [],
                "env": {}
              },
              "hub-tooling/dummy-tools": {
                "attributes": {
                  "memoryLimitBytes": "1073741824"
                },
                "servers": {},
                "volumes": {
                  "projects": {
                    "path": "/projects"
                  }
                },
                "installers": [],
                "env": {}
              }
            },
            "recipe": {
              "type": "kubernetes",
              "content": "kind: List\nitems:\n - \n  apiVersion: v1\n  kind: Pod\n  metadata:\n   name: hub-tooling\n  spec:\n   volumes:\n    - \n     name: cfg\n     emptyDir: {}\n   initContainers:\n    - \n     image: 'busybox:latest'\n     name: attach-file\n     command: ['/bin/sh', '-c', 'echo abcdef > /cfg/xyzFile']\n     resources:\n      limits:\n       memory: 512Mi\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n   containers:\n    - \n     image: 'eclipse/che-theia:latest'\n     name: hub-theia-ide\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: 'busybox:latest'\n     name: dummy-tools\n     command: ['/bin/sh', '-c', '[ -f /cfg/xyzFile ] && cp /cfg/xyzFile /xyzFile; tail -f /dev/null']\n     volumeMounts:\n      - \n       name: cfg\n       mountPath: /cfg\n     resources:\n      limits:\n       memory: 1024Mi\n    - \n     image: 'wsskeleton/che-machine-exec:latest'\n     name: machine-exec\n     resources:\n      limits:\n       memory: 512Mi\n",
              "contentType": "text/x-yaml"
            }
          }
        },
        "projects": [],
        "name": "default",
        "attributes": {},
        "commands": [],
        "links": []
      },
      "description": "Dummy Tooling",
      "creator": "Dummy Software",
      "scope": "general",
      "links": []
    }

```

So the above stack will reproduce the issue. It works if we remove the initContainer part completely, otherwise it does not.

**Comment by vtawniya** [comment]
@eivantsov 
I should have mentioned the version first. We are using eclipse che 6.14.

**Comment by VikTawniya** [comment]
Issue was with initContainers in 6.14 version, for which I have raised a separate bug.
